### PR TITLE
Migration to dotnet core cli

### PR DESCRIPTION
**Pull request by joaope**
Hi,

Hope it's fine for me to do this PR. Nice piece of software that deserves a bit of work :)

Few points:
- It required some work at code level. Some things significantly changed with Owin on ASP.NET core latest release. Mainly around middleware implementation and execution in the pipeline.
- There's no `WebClient` anymore on core. Most unit tests changes were around that, replacing it with `HttpClient`.
- Removed a bunch of packaging-related files as _dotnet pack_ takes care of the hard work together with _project.json_ files. There were some ncrunch files lying around as well. Unfortunately, it's not supported yet because tooling for core is a bit behind schedule.
- I can see you guys use appveyor. Added the .yml file to the source. If you want to use it it requires a few bits and bobs around deploy.
- Tests can be executed with _dotnet test_. Visual Studio 2015 also supports xUnit on its Tests Explorer (from Update 2 on).

I tried to maintain code compatibility from previous version. Think I did it.

### CODE REVIEW

File: JustFakeIt/project.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            "buildOptions": {
+                "optimize": true
+            }
+        }
+    },
+    "frameworks": {
+        "netstandard13": {
+            "dependencies": {
+                "Microsoft.AspNetCore.Http.Abstractions": "1.0.0",
+                "Microsoft.AspNetCore.Owin": "1.0.0",
```
Comment (slang25) [inline JustFakeIt/project.json]: You can remove this reference as there is no longer any owin dependancy, this package is for interop/bridge for owin.

Comment (slang25) [inline JustFakeIt/project.json]: See [the docs](https://docs.asp.net/en/latest/fundamentals/owin.html#running-owin-middleware-in-the-asp-net-pipeline)

**Comment by slang25** [comment]
@joaope This PR is excellent 😄 I've one very minor comment that I'd like addressed, once this is addressed I am happy to merge (unless someone else spots a show stopper), then I'll sort out appveyor later in the week. Thanks for your contribution.

**Comment by slang25** [comment]
One last request, could you delete the .nuget folder in the root. It's no longer needed, and has mention of an older version of xunit.

**Comment by AnthonySteele** [comment]
:+1: Core all the things (eventually). I assume that it will still work on .net crust (i.e. not corre)

**Comment by slang25** [comment]
It works on full 4.6 and up as it is .NET Standard 1.3. Although for me I had a bug where I needed to copy the native libuv dll to the bin folder, this was an old bug that should have been fixed, I don't trust my setup here.

**Comment by slang25** [comment]
Thinking about my last post, it probably is not preferable on a windows machine to have a native dependency for tests, you might need to wrestle with config and project files for tests runners, NCrunch, VS, R# etc...

It might be best to not include a server dependency, but have it configurable and require the user to include it, using:

```.UseServer("Microsoft.AspNet.Server.Kestrel") ```
or
```.UseServer("Microsoft.AspNet.Server.WebListener") ```

depending on what's configured. Maybe even have this determined from loaded assemblies, but still with an option to make it explicit.

This way you have the best of all worlds, cross-plat, core, and no native windows deps.

I don't think that blocks this PR, but certainly would be something I'd like to see.

**Comment by slang25** [comment]
Another option, target both .NET 4.5.1 and .NET Standard 1.3, all of the AspNetCore packages (at a quick glance) have a version for 4.5.1, the helps lower the version requirement from 4.6, and we can include kestrel in the core version, and web listener (in preview, sue me) in the full framework one.

File: JustFakeIt/project.json (hunk 1 above)
Comment (joaope) [inline JustFakeIt/project.json]: Left overs from when I was trying to understand Owin to AspNetCore changes. Removed.

**Comment by joaope** [comment]
I totally thought the same about the server but with different conclusions: to be honest I think it's fine to use Kestrel-only solution. Giving the user the ability to specify the server has other repercussions, including him/her to start playing too much with the server implementation itself which is a bit out of the lib scope. Kestrel is (and will always be) cross-platform. That way the lib has the freedom to play with the server, not the end-user.

As with every library, it's preferable to constraint from the start and slowly open it's capabilities than to give everything away from day one and then had to support it for the time being :)

That being said, I think you guys are right, we need to make an effort on lowering framework support to netstandard12 (net 4.5.1). And do it on this PR probably, to be like that since day one.

**Comment by joaope** [comment]
_Another option, target both .NET 4.5.1 and .NET Standard 1.3, all of the AspNetCore packages (at a quick glance) have a version for 4.5.1, the helps lower the version requirement from 4.6, and we can include kestrel in the core version, and web listener (in preview, sue me) in the full framework one._

I like this option!

**Comment by slang25** [comment]
@joaope cool, I was hoping you'd like that, cause it seems like it keeps all people happy. If you are on full framework, you want it to "just work", you don't want a native dependency, you don't want to be forced to upgrade framework. For core, you want kestrel, you want cross plat and all that goodness.

I've attempted the change locally, and it's only a minor change, but I'm being stupid cause it's complaining at runtime it can't load ```System.Runtime```, I am a failure. Would you try it and let me know if you get better success?

**Comment by joaope** [comment]
Turns out the problem is a bit different. Kestrel is supported on framework 4.5.1.

Funny enough it's `WebHeaderCollection` which is not :). It's no longer on _System.dll_ but _System.Net.WebHeaderCollection.dll_. Even the implementation changed but not significantly enough for me to have to change any code.

Check my last commit.

**Comment by slang25** [comment]
I got mine working in the end, didn't notice a rogue assembly binding redirect. So I can get it to build and run on both platforms with:
```
  "dependencies": {
    "Microsoft.AspNetCore.Http.Abstractions": "1.0.0",
    "Newtonsoft.Json": "9.0.1"
  },
  "frameworks": {
    "netstandard13": {
      "dependencies": {
        "Microsoft.AspNetCore.Server.Kestrel": "1.0.0",
        "NETStandard.Library": "1.6.0",
        "System.Collections.Specialized": "4.0.1",
        "System.Diagnostics.TraceSource": "4.0.0",
        "System.Linq": "4.1.0",
        "System.Net.NetworkInformation": "4.1.0",
        "System.Net.WebHeaderCollection": "4.0.1"
      }
    },
    "net451": {
      "dependencies": {
        "Microsoft.AspNetCore.Server.WebListener": "0.1.0"
      }
    }
  }
```
and
```
            var host = new WebHostBuilder()
                .UseUrls(BaseUri.AbsoluteUri)
                .Configure(app => app.UseProxyMiddleware(Expect, capturedRequests))
#if NET451
                .UseWebListener()
#elif NETSTANDARD1_3
                .UseKestrel()
#endif
                .Build();
```

Although in honesty, I don't know what I'm doing, for example what goes in "dependencies" vs "frameworkAssemblies" for example.

**Comment by slang25** [comment]
Ah cool, looking at your last commit yours is the same as mine with the unnecessary cruft removed. If you move kestrel to netstandard13 and add weblistener to net451, then add the ifdef we are all good to merge I think 😄

**Comment by joaope** [comment]
I'm afraid I'm not following you. Why do you want to use two different servers? Kestrel is cross-platform while WebListener is not. Your approach is effectively making the net451 version of the library to work only on windows.

**Comment by slang25** [comment]
Yes, I'm suggesting that if you are targeting core then you want kestrel cause it's cross-plat. But for .NET 4.5.1 people who do a nuget update will likely be on windows and will probably benefit from not having a native dependency. A sweeping statement I know, but I think it's right approach.

**Comment by slang25** [comment]
There could be people using mono on linux/osx, and want to upgrade, and want to stay on full framework, I would say this is relatively edge case. We could provide some sort of api to override the behavior for them.

**Comment by joaope** [comment]
Everybody is using nuget now, not only windows/mono users. As it is now it will work perfectly on mono. Not sure what you mean by native dependency. _Microsoft.AspNetCore.Server.Kestrel_ is a nuget dll package that supports both .net4.5.1 and netstandard13+. Exactly like _Microsoft.AspNetCore.Server.WebListener_.

I totally understand that support to inject a custom server might be cool, but it has not much to do it Kestrel vs. WebListener.

**Comment by slang25** [comment]
I didn't word that very well above, I did not mean people who use nuget will be on windows, I mean people using this library on 4.5.1 will likely be on windows, it's just more popular.
By native dependency I mean it ships with libuv, so libuv.dll needs to be present in the bin folder (or fusion probing paths or whatever), whereas WebListener does not ship with anything other than managed code assemblies, it does do native windows calls however, but I'm not counting that.

My experience in the past is that when running unit tests (and that's one of the primary use cases for JustFakeIt) using test runners such as NCrunch and R# test runner they sometimes do shadow copying of assemblies pr various other optimization tricks, and do not copy over the native assemblies, you can work around these issues but it is a pain.

The situation may have changed since then (a year or 2 ago now). I wouldn't want someone to upgrade and to experience pain at our expense. I'll talk to some colleagues tomorrow, this might just me and it's not an issue.

**Comment by joaope** [comment]
Btw:

* _dependencies_ outside _frameworks_ will be common to all specified frameworks packages
* _dependencies_ inside _each framework_ are specific for that framework package
* _frameworkAssemblies_ is for you to be able to reference .net framework dlls. Remember the time when you had not only full framework installers but partial ones, where not all dlls are available. Not only that but, for example, Portable Class Library (PCL). So you will usually only use _frameworkAssemblies_ when targeting _netXXX_ frameworks.

mscorlib is referenced by default on _netXXX_. That's why, for example, _System.Net.NetworkInformation_ and _System.Net.WebHeaderCollection_ only need to be referenced by netstandard13, because on .net full framework they are part of _System.dll_ and consequently part of mscorlib.

Moreover, that's why on netstandard/netcoreapp you specify every single dll. Because with core that's what they are in practice, a nuget package dll. Dll all the things!

.NET core is not easy to grasp and that abstraction layer of _netstandard_ on top of everything, although a brilliant idea, introduced some confusion. It will get some time till people to fully understand how all pieces work together. Nice stuff tho! :)

**Comment by slang25** [comment]
Thanks for that, that's made it clearer 👍

**Comment by joaope** [comment]
Oh! I see what you mean. I don't know how it was before but look, [libuv.dll is on nuget now](https://www.nuget.org/packages/Libuv/) and it's referenced by [Microsoft.AspNetCore.Server.Kestrel](https://www.nuget.org/packages/Microsoft.AspNetCore.Server.Kestrel/). It's just a dependency like the rest of them.

**Comment by slang25** [comment]
True, but nuget is just still the deployment mechanism, at some point these dlls land in a folder and get run, if it as native dll then it has to be the right bitness and it gets loaded by a different mechanism, this has been the source of problems for me in the past, but you know what, I'll just try it out rather than speculate.

**Comment by slang25** [comment]
@joaope just a quick update, just in case you thought I was sitting on this. I've started to test this version with some of our internal systems to see if there are any major issues, there appears to be a weird bug where paths with query strings end up being interpenetrated in the middleware as having ```??``` instead of ```?```, which is really strange given that there is a passing unit test with a query string in it in this solution.

I'll do some more investigation and keep you updating on the status.

**Comment by joaope** [comment]
hey @slang25, tks for the update. Will check it out as well if I have the time, later today.

**Pull request closed without merging**